lib/chef/resource/macos_userdefaults.rb

```diff
@@ -73,54 +107,90 @@ def coerce_booleans(val)
+          Chef::Log.debug "#load_current_value: #{state_cmd.join(" ")} returned stdout: #{state.stdout} and stderr: #{state.stderr}"
+          current_value_does_not_exist!
+        end
+
+        # parse the output from the defaults command to ruby data type
+        # todo: This is a pretty basic implementation. PRs welcome ;)
+        case state.stdout[0]
+        when "{" # dict aka hash
+          # https://rubular.com/r/cBnFu1nttMdsXq
+          data = /^\s{2,}(.*)\s=\s(.*);/.match(state.stdout)
+          fail_if_unparsable(data, state.stdout)
+          value Hash[*data.captures]
+        when "(" # array
+          # https://rubular.com/r/TfYejXUJny11OG
+          data = /^\s{2,}(.*),?/.match(state.stdout)
+          fail_if_unparsable(data, state.stdout)
+          value data.captures
+        else # a string/int/float/bool
+          value state.stdout.strip
+        end
+      end
+
+      #
+      # If there were not matches raise a warning that we couldn't parse the output of the defaults
+      # CLI and return a nil current_resource by calling current_value_does_not_exist!
+      #
+      # @param [MatchData] match_data
+      # @param [String] stdout
+      #
+      def fail_if_unparsable(match_data, stdout)
+        return unless match_data.captures.empty?
+
+        Chef::Log.warn("Could not parse macos defaults CLI value data: #{stdout}.")
+        current_value_does_not_exist!
       end
 
       action :write do
-        description "Write the setting to the specified domain"
+        description "Write the value to the specified domain/key."
 
-        unless current_resource.is_set
-          cmd = ["defaults write"]
-          cmd.unshift("sudo") if new_resource.sudo
+        converge_if_changed do
+          cmd = defaults_write_cmd
+          Chef::Log.debug("Updating defaults value by shelling out: #{cmd.join(" ")}")
 
-          cmd << if new_resource.global
-                   "NSGlobalDomain"
-                 else
-                   "'#{new_resource.domain}'"
-                 end
+          if new_resource.user.nil?
+            shell_out!(cmd)
+          else
+            shell_out!(cmd, user: new_resource.user)
+          end
+        end
+      end
 
-          cmd << "'#{new_resource.key}'" if new_resource.key
+      action_class do
+        def defaults_write_cmd
           value = new_resource.value
-          type = new_resource.type.empty? ? value_type(value) : new_resource.type
-          # creates a string of Key1 Value1 Key2 Value2...
-          value = value.map { |k, v| "\"#{k}\" \"#{v}\"" }.join(" ") if type == "dict"
-          if type == "array"
-            value = value.join("' '")
-            value = "'#{value}'"
-          end
+          type = new_resource.type || value_type(value)
+
+          value = case type
+                  when "dict"
+                    # creates an array of Key1 Value1 Key2 Value2...
+                    value.flatten.map { |x| "'#{x}'" }
```

Comment:
This comment isn't quite true, it takes each one and makes a string with single quote around it. That comment would be more accurate saying `creates and array of ["'key1'", "'key2'"]` 

```
irb(main):003:0> d = {'ast' => 'dink'}
irb(main):004:0> d.flatten.map { |x| "'#{x}'" }
=> ["'ast'", "'dink'"]
```

Did you instead want `.to_a.flatten`?

```
irb(main):006:0> d.to_a.flatten
=> ["ast", "dink"]
```

Reply:
Fair point. Updated

Reply:
The values need to be quoted as they go out to the command so that they don't fail if they have a space in them. That's one of the reported bugs that was fixed here and the test case added for our azure pipelines job.